packages/core/saved-objects/core-saved-objects-utils-server/src/merge_migrations.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { ary, isFunction, mergeWith } from 'lodash';
+import type {
+  SavedObjectMigrationContext,
+  SavedObjectMigration,
+  SavedObjectMigrationMap,
+  SavedObjectUnsanitizedDoc,
+} from '@kbn/core-saved-objects-server';
+
+/**
+ * Composes two migrations into a single migration.
+ * ```
+ * mergeSavedObjectMigrations(outer, inner) -> (doc, context) => outer(inner(doc, context), context) }
+ * ```
+ *
+ * If at least one of the migrations is not deferred, the composed migration will not be deferred.
+ *
+ * @public
+ * @param outer Wrapping migration.
+ * @param inner Wrapped migration.
+ * @param rest Additional wrapped migrations to compose.
+ * @returns The composed migration can be either a function or an object depending on the input migrations.
+ */
+export function mergeSavedObjectMigrations(
+  outer: SavedObjectMigration,
+  inner: SavedObjectMigration,
+  ...rest: SavedObjectMigration[]
+): SavedObjectMigration {
+  if (rest.length) {
+    return mergeSavedObjectMigrations(
+      mergeSavedObjectMigrations(outer, inner),
+      ...(rest as [SavedObjectMigration, ...SavedObjectMigration[]])
+    );
+  }
+  if (!inner || !outer) {
+    return inner || outer;
+  }
+
+  const merged = mergeWith(
+    { ...(isFunction(inner) ? { transform: inner } : inner) },
+    isFunction(outer) ? { transform: outer } : outer,
+    (innerValue, outerValue, key) => {
+      if (key === 'deferred') {
+        return !!(innerValue && outerValue);
+      }
+
+      if (key === 'transform') {
+        return (state: SavedObjectUnsanitizedDoc, context: SavedObjectMigrationContext) =>
+          outerValue(innerValue(state, context), context);
+      }
+
+      return inner ?? outer;
+    }
+  );
+
+  if (isFunction(inner) && isFunction(outer)) {
+    return merged.transform;
+  }
+
+  return merged;
+}
+
+/**
+ * Merges two saved object migration maps.
+ *
+ * If there is a migration for a given version on only one of the maps,
+ * that migration function will be used:
+ *
+ * mergeSavedObjectMigrationMaps({ '1.2.3': f }, { '4.5.6': g }) -> { '1.2.3': f, '4.5.6': g }
+ *
+ * If there is a migration for a given version on both maps, the migrations will be composed:
```

Comment:
We're relying on `lodash` to match the version here. Are we ok with that or should we add our own validation on version match?

Reply:
That's acceptable since we are versioning migrations using the Kibana version, so that exact match is good enough here.